include/tc/core/cuda.h

```diff
@@ -84,7 +84,31 @@ struct WithDevice {
   size_t newGpu;
 };
 
-// Query the active device about the avaialble memory size.
-size_t querySharedMemorySize();
+//
+// This functionality in this type of class has been rewritten over and over
+// again. Here we just provide a static singleton and basic properties.
+// Consider lifting stuff up from fbcuda rather than reinventing the wheel
+//
+class CudaGPUInfo {
+  CudaGPUInfo(
+      const std::vector<std::string>& gpuNames,
+      const std::vector<size_t>& sharedMemSizes)
+      : gpuNames_(gpuNames), sharedMemSizes_(sharedMemSizes) {}
+
+ public:
+  static CudaGPUInfo& GPUInfo();
+
+  // These functions require init to have been run, they are thus members of
+  // the singleton object and not static functions.
+  int NumberGPUs() const;
```

Comment:
2 non-blocking remarks here:
1. querying the cuda runtime API should generally be cached as much as possible because calls like ```cudaGetDeviceProperties``` may add up to non-trivial overheads. We should measure it first but we need to be aware of it, see e.g. [here](https://github.com/facebook/fbcuda/blob/8ef2af9b579b8610c59ab0ccf6e9075350de4320/util/CachedDeviceProperties.cpp) 
2.  nit ```lowerCamelCase```

Reply:
this is merely moving the existing code around, will fix later